core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -562,44 +513,60 @@ public IndicesOptions getOptions() {
-                    indices = metaData.concreteAllIndices();
-                } else if (options.expandWildcardsOpen()) {
-                    indices = metaData.concreteAllOpenIndices();
-                } else if (options.expandWildcardsClosed()) {
-                    indices = metaData.concreteAllClosedIndices();
+
+                final IndexMetaData.State excludeState;
+                if (options.expandWildcardsOpen() && options.expandWildcardsClosed()){
+                    excludeState = null;
+                } else if (options.expandWildcardsOpen() && options.expandWildcardsClosed() == false) {
+                    excludeState = IndexMetaData.State.CLOSE;
+                } else if (options.expandWildcardsClosed() && options.expandWildcardsOpen() == false) {
+                    excludeState = IndexMetaData.State.OPEN;
                 } else {
                     assert false : "this shouldn't get called if wildcards expand to none";
-                    indices = Strings.EMPTY_ARRAY;
+                    excludeState = null;
                 }
-                boolean found = false;
-                // iterating over all concrete indices and see if there is a wildcard match
-                for (String index : indices) {
-                    if (Regex.simpleMatch(aliasOrIndex, index)) {
-                        found = true;
-                        if (add) {
-                            result.add(index);
-                        } else {
-                            result.remove(index);
+
+                final Map<String, AliasOrIndex> matches;
+                if (Regex.isMatchAllPattern(expression)) {
+                    // Can only happen if the expressions was initially: '-*'
+                    matches = metaData.getAliasAndIndexLookup();
+                } else if (expression.endsWith("*")) {
+                    // Suffix wildcard:
+                    assert expression.length() >= 2 : "expression [" + expression + "] should have at least a length of 2";
+                    String fromPrefix = expression.substring(0, expression.length() - 1);
+                    char[] toPrefixCharArr = fromPrefix.toCharArray();
+                    toPrefixCharArr[toPrefixCharArr.length - 1]++;
+                    String toPrefix = new String(toPrefixCharArr);
+                    matches = metaData.getAliasAndIndexLookup().subMap(fromPrefix, toPrefix);
+                } else {
+                    // Other wildcard expressions:
+                    final String pattern = expression;
+                    matches = filterEntries(metaData.getAliasAndIndexLookup(), new Predicate<Map.Entry<String, AliasOrIndex>>() {
+                        @Override
+                        public boolean apply(@Nullable Map.Entry<String, AliasOrIndex> input) {
+                            return Regex.simpleMatch(pattern, input.getKey());
                         }
-                    }
+                    });
```

Comment:
Before it would evaluate all all indices and aliases and check if it matches with the current expression being evaluated.
Now functionally it does the same, but it tries to be smarter and use the SortedMap is possible.

It tries to generate a list of candidate matches:
1) `-*` expression, we can't be smart here, so all aliases and indices need to be evaluated. (`*` is already covered in the beginning of this method)
2) Suffix wildcards we can use SortedMap#subMap() to quickly lookup the candidates matching with the wildcard expressions.
3) All other wildcard expressions (`*` in the beginning or middle) we need to fall back to old behaviour and just check  everything.

Then for all the candidate matches it checks based on the indices options and whether the index state is closed or open (or the index the alias it pointing to). To make a final decision to include or exclude.